opencog/reasoning/RuleEngine/rule-engine-src/pln/DefaultForwardChainerCB.cc

```diff
@@ -163,22 +164,67 @@ HandleSeq DefaultForwardChainerCB::get_rootlinks(Handle htarget, AtomSpace* as,
 }
 HandleSeq DefaultForwardChainerCB::choose_input(FCMemory& fcmem)
 {
+    HandleSeq inputs;
+    PLNCommons pc(as_);
     Handle htarget = fcmem.get_cur_target();
-    //get everything associated with the target handle
-    HandleSeq inputs = as_->getNeighbors(htarget, true, true, LINK, true);
+
+    //Get everything associated with the target handle.
+    HandleSeq neighbors = as_->getNeighbors(htarget, true, true, LINK, true);
+
+    //Add all root links of atoms in @param neighbors.
+    for (auto hn : neighbors) {
+        if (hn->getType() != VARIABLE_NODE) {
+            HandleSeq roots;
+            pc.get_root_links(hn, roots);
+            for (auto r : roots) {
+                if (find(inputs.begin(), inputs.end(), r) == inputs.end() and r->getType()
+                        != BIND_LINK)
+                    inputs.push_back(r);
+            }
+        }
+    }
+
     return inputs;
 }
 
 Handle DefaultForwardChainerCB::choose_next_target(FCMemory& fcmem)
 {
-    HandleSeq tlist = fcmem.get_target_list();
+    HandleSeq tlist = fcmem.get_premise_list();
```

Comment:
So are you renaming "target" to "premise"?  How come this method is still "choose_next_target"?